tinybmp/src/lib.rs

```diff
@@ -1,251 +1,243 @@
-        u32::from(self.header.bpp)
+    /// Returns an iterator over the raw pixels in the image.
+    ///
+    /// The iterator returns the raw pixel colors as `u32` values. To automatically convert the raw
+    /// values into the color specified by `C` use [`pixels`] instead.
+    ///
+    /// [`pixels`]: #method.pixels
+    pub fn raw_pixels<'b>(&'b self) -> RawPixels<'b, 'a, C> {
+        RawPixels::new(self)
     }
 
     /// Returns the row length in bytes.
     ///
     /// Each row in a BMP file is a multiple of 4 bytes long.
     fn bytes_per_row(&self) -> usize {
-        let bits_per_row = self.width() as usize * self.bpp() as usize;
+        let bits_per_row = self.header.image_width as usize * usize::from(self.header.bpp.bits());
 
         (bits_per_row + 31) / 32 * (32 / 8)
     }
 }
 
-impl<'a> IntoIterator for &'a BmpRaw<'a> {
-    type Item = Pixel;
-    type IntoIter = BmpIterator<'a>;
-
-    fn into_iter(self) -> Self::IntoIter {
-        let pixel_stride = match self.bpp() {
-            1 | 8 | 16 | 24 | 32 => self.bpp() as usize,
-            depth => panic!("Bit depth {} not supported", depth),
-        };
-
-        BmpIterator {
-            bmp: self,
-            pixel_data: self.image_data(),
-            pixel_stride,
-            x: 0,
-            y: 0,
-            bit_idx: 0,
-        }
+impl<'a> Bmp<'a, ()> {
+    /// Create a bitmap object from a byte slice.
+    ///
+    /// The created object keeps a shared reference to the input and does not dynamically allocate
+    /// memory.
+    ///
+    /// In contrast to the [`from_slice`] constructor no color type needs to be specified when
+    /// calling this method. This will disable all functions that requires a specified color type,
+    /// like the [`pixels`] method.
+    ///
+    /// [`from_slice`]: #method.from_slice
+    /// [`pixels`]: #method.pixels
+    pub fn from_slice_raw(bytes: &'a [u8]) -> Result<Self, ParseError> {
+        Self::from_slice_common(bytes)
     }
 }
 
-/// Iterator over individual BMP pixels
-///
-/// Each pixel is returned as a `u32` regardless of the bit depth of the source image.
-#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug)]
-pub struct BmpIterator<'a> {
-    /// Reference to original BMP image
-    bmp: &'a BmpRaw<'a>,
-
-    /// Image pixel data as a byte slice, little endian ordering
-    pixel_data: &'a [u8],
-
-    /// Number of bits per pixel
-    pixel_stride: usize,
-
-    /// Current X position
-    x: u32,
+impl<'a, C> Bmp<'a, C>
+where
+    C: PixelColor,
+{
+    /// Create a bitmap object from a byte slice.
+    ///
+    /// The created object keeps a shared reference to the input and does not dynamically allocate
+    /// memory.
+    ///
+    /// The color type needs to be specified explicitly when this method is called, for example by
+    /// using the turbofish syntax. An error is returned if the bit depth of the specified color
+    /// type doesn't match the bit depth of the BMP file.
```

Comment:
As above, might it be worth adding something like "to load an image without e-g support/integration, use `from_slice_raw`?

It isn't immediately clear what the difference between these two methods is when looking at the rendered docs, hence the suggestion that naming e-g as the reason for these _two_ methods to exist.

Reply:
The docs need some work, but I would prefer to add this to the PR for #424. That PR will add additional ways of using `tinybmp` and it will be easier to write an overview of the different variants when everything is implemented.

Reply:
Sure, that makes sense.